lib/stm32/fdcan.c

```diff
@@ -0,0 +1,891 @@
+
+		if (pending_frames)
+		{
+			FDCAN_RXF1A(canport) |= (get_index & FDCAN_RXF1A_R1AI_MASK) 
+				<< FDCAN_RXF1A_R1AI_SHIFT;
+		}
+	}
+	else
+	{
+		return FDCAN_E_INVALID;
+	}
+
+	if (pending_frames == 0)
+	{
+		return FDCAN_E_NOTAVAIL;
+	}
+
+	return FDCAN_E_OK;
+}
+
+/** Enable IRQ from FDCAN block.
+ * This routine configures FDCAN to enable certain IRQ.
+ * Each FDCAN block supports two IRQs.
+ * @param canport FDCAN block base address
+ * @param irq number of IRQ to be enabled (currently 0 or 1)
+ */
+void fdcan_enable_irq(uint32_t canport, uint32_t irq)
+{
+	FDCAN_ILE(canport) |= irq & (FDCAN_ILE_INT0 | FDCAN_ILE_INT1);
+}
+
+/** Disable IRQ from FDCAN block.
+ * This routine configures FDCAN to disable certain IRQ.
+ * Each FDCAN block supports two IRQs.
+ * @param canport FDCAN block base address
+ * @param irq number of IRQ to be enabled (currently 0 or 1)
+ */
+void fdcan_disable_irq(uint32_t canport, uint32_t irq)
+{
+	FDCAN_ILE(canport) &= ~(irq & (FDCAN_ILE_INT0 | FDCAN_ILE_INT1));
+}
+
+/** Check if there is free transmit buffer.
+ * @param canport FDCAN port
+ * @returns true if there is at least one free transmit buffer for new message
+ * to be sent, false otherwise.
+ */
+bool fdcan_available_tx(uint32_t canport)
+{
+	return (fdcan_get_free_txbuf(canport) != FDCAN_E_BUSY);
+}
+
+/** Tell if there is message waiting in receive FIFO.
+ * @param canport FDCAN port
+ * @param fifo Rx FIFO number, 0 or 1
+ * @returns true if there is at least one message waiting in given receive FIFO,
+ * false otherwise.
+ */
+bool fdcan_available_rx(uint32_t canport, uint8_t fifo)
+{
+	unsigned get_index, pending_frames;
+	if (fdcan_get_fill_rxfifo(canport, fifo, &get_index, &pending_frames) != FDCAN_E_OK)
+	{
+		return false;
+	}
+
+	return (pending_frames != 0);
+}
+
+/**@}*/
+
+/* --- FD-CAN internal functions -------------------------------------------- */
+
+/** Routine implementing FDCAN_CCCR's INIT bit manipulation.
+ * This routine will change INIT bit and wait for it to actually
+ * change its value. If change won't happen before timeout,
+ * error is signalized. If INIT bit already has value which
+ * should be set, this function will return immediately.
+ * @param canport FDCAN block base address
+ * @param set new value of INIT, true means set
+ * @returns FDCAN_E_OK on success, FDCAN_E_TIMEOUT if INIT bit value
+ * didn't change before timeout has expired.
+ */
+static int fdcan_cccr_init_cfg(uint32_t canport, bool set)
+{
+	uint32_t expected;
+	uint32_t wait_ack;
+
+	if (set)
+	{
+		if ((FDCAN_CCCR(canport) & FDCAN_CCCR_INIT) == FDCAN_CCCR_INIT)
+		{
+			/* Already there, sir */
+			return FDCAN_E_OK;
+		}
+
+		FDCAN_CCCR(canport) |= FDCAN_CCCR_INIT;
+		expected = FDCAN_CCCR_INIT;
+	}
+	else
+	{
+		if ((FDCAN_CCCR(canport) & FDCAN_CCCR_INIT) == 0)
+		{
+			/* Already there, sir */
+			return FDCAN_E_OK;
+		}
+
+		FDCAN_CCCR(canport) &= ~FDCAN_CCCR_INIT;
+		expected = 0;
+	}
+
+	/* Wait, until INIT bit is acknowledged */
```

Comment:
if you split this check into an _is_ok, or _is_init_complete or something, then you can have a function that sets and returns, and the users then polls _is_ok() themselves, and also an _init_blocking() that does the default